src/rpc/methods/eth/filter/mod.rs

```diff
@@ -186,6 +195,178 @@ impl EthEventHandler {
             Ok(false)
         }
     }
+
+    fn parse_eth_filter_spec<DB: Blockstore>(
+        &self,
+        ctx: &Ctx<DB>,
+        filter_spec: &EthFilterSpec,
+    ) -> anyhow::Result<ParsedFilter> {
+        EthFilterSpec::parse_eth_filter_spec(
+            filter_spec,
+            ctx.chain_store().heaviest_tipset().epoch(),
+            self.max_filter_height_range,
+        )
+    }
+
+    async fn collect_events<DB: Blockstore + Send + Sync + 'static>(
+        &self,
+        ctx: &Ctx<DB>,
+        tipset: &Arc<Tipset>,
+        spec: &EthFilterSpec,
+        collected_events: &mut Vec<CollectedEvent>,
+    ) -> anyhow::Result<()> {
+        let tipset_key = tipset.key().clone();
+        let height = tipset.epoch();
+
+        let messages = ctx.chain_store().messages_for_tipset(tipset)?;
+
+        let StateOutput { events, .. } = ctx.state_manager.tipset_state_plus(tipset).await?;
+
+        let events = if let Some(events) = events {
+            events
+        } else {
+            tracing::warn!("No events stored");
+            return Ok(());
+        };
+        for (i, (message, events)) in messages.iter().zip(events.into_iter()).enumerate() {
+            for (j, event) in events.iter().enumerate() {
+                let id_addr = Address::new_id(event.emitter());
+                let result = ctx
+                    .state_manager
+                    .resolve_to_deterministic_address(id_addr, tipset.clone())
+                    .await
+                    .with_context(|| {
+                        format!(
+                            "resolving address {} failed (EPOCH = {})",
+                            id_addr,
+                            tipset.epoch()
+                        )
+                    });
+                if result.is_err() {
+                    // Skip event
+                    continue;
+                };
+                let resolved = result.expect("Infallible");
```

Comment:
nit: might be less scary and no need to check out the previous chunk of code to be sure it doesn't panic.

```suggestion
                let Ok(resolved) = result {
                    resolved
                } else {
                    continue;
                }
```